File: bioprocs/scripts/utils/mem2.py
if 'mem2' not in vars() or not callable (mem2):
	
	def _autoUnit(num):
		if num % (1024 * 1024) == 0:
			return num / (1024*1024), 'G'
		elif num % 1024 == 0:
			return num / 1024, 'M'
		else:
			return num, 'K'

	# unit = 'auto'/'G'/'M'/'K'/'java'
	def mem2 (mem, unit = 'auto'):
		mem   = str(mem)
		ounit = mem[-1].upper()
		if ounit == 'G':
			num   = int(mem[:-1]) * 1024 * 1024
		elif ounit == 'M':
			num   = int(mem[:-1]) * 1024
		elif ounit == 'K':
			num   = int(mem)
		elif not ounit.isdigit():
			raise ValueError('Unknown memory unit: ' + ounit)
		else:
			num   = int(mem)
		
		unit = unit.upper()
		retn, retu = _autoUnit(num)
		if unit == 'G':
			if retu == 'M':   retn /= 1024.
			elif retu == 'K': retn /= (1024. * 1024.)
			retu = 'G'
		elif unit == 'M':
			if retu == 'G':   retn *= 1024
			elif retu == 'K': ret /= 1024.
			retu = 'M'
		elif unit == 'K':
			if retu == 'G':   retn *= 1024 * 1024
			elif retu == 'M': retn *= 1024 
			retu = 'K'
		
		if unit == 'JAVA':
			xmx = "-Xmx" + str(retn) + retu
			n, u = _autoUnit(num / 8)
			return '-Xms' + str(n) + u + ' ' + xmx
		else:
			return str(retn) + retu

File: bioprocs/scripts/utils/mem2.r
if (!exists('mem2')) {

	autoUnit = function (num) {
		if (num %% (1024 * 1024) == 0) {
			return(list(n = num / 1024 / 1024, u = 'G'))
		} else if (num %% 1024 == 0) {
			return(list(n = num / 1024, u = 'M'))
		} else {
			return(list(n = num, u = 'K'))
		}
	}

	mem2 = function (mem, unit = 'auto') {

		if (grepl('G$', mem) || grepl('g$', mem)) {
			num = as.numeric (substr(mem, 1, nchar(mem)-1)) * 1024 * 1024
		} else if (grepl('M$', mem) || grepl('m$', mem)) {
			num = as.numeric (substr(mem, 1, nchar(mem)-1)) * 1024
		} else if (grepl('K$', mem) || grepl('k$', mem)) {
			num = as.numeric (substr(mem, 1, nchar(mem)-1))
		} else {
			num = as.numeric (mem)
		}

		ret = autoUnit(num)
		if (unit == 'G' || unit == 'g') {
			if (ret$u == 'M') ret$n = ret$n / 1024.0
			if (ret$u == 'K') ret$n = ret$n / 1024.0 / 1024.0
			ret$u = 'G'
		} else if (unit == 'M' || unit == 'm') {
			if (ret$u == 'G') ret$n = ret$n * 1024
			if (ret$u == 'K') ret$n = ret$n / 1024.0 
			ret$u = 'M'
		} else if (unit == 'K' || unit == 'k') {
			if (ret$u == 'G') ret$n = ret$n * 1024.0 * 1024.0
			if (ret$u == 'M') ret$n = ret$n * 1024.0
			ret$u = 'K'
		}

		if (unit == 'java' || unit == 'Java' || unit == 'JAVA') {
			xmx = paste0("-Xmx", ret$n, ret$u)
			r   = autoUnit(num / 8)
			return(paste(paste0('-Xms', r$n, r$u), xmx))
		} else {
			return(paste0(ret$n, ret$u))
		}
	}
}